_drafts/2024-09-01-aws-terraform-02.md

```diff
@@ -0,0 +1,272 @@
+
+```
+We'll also need to enable read/write/execute permissions on the script in order to run it.
+
+`chmod 700 build-and-push.sh`  
+
+Now run the script and if all went well you should see the following output:
+
+
+![console output from pushing an image container to ecr](/assets/getting-started-with-aws-tf-02/example-push-ecr-output.png)
+
+
+## Step 3 - Pull the ECR image onto EC2 instance and run it
+
+By now we should have an EC2 instance running in AWS and a container image hosted on ECR. Now we just need to run the image on our EC2 instance.
+
+In order to achieve this we'll need to:
+1. Add a security egress rule to allow EC2 to connect to ECR
+2. Define a role to allow read access to ECR and assign it to our EC2 instance
+3. write another custom script to run on our instance which will pull the docker image and run it in its own Docker environment 
+
+Lets define our security group rule first.
+
+Open the `unicorn-api.tf` file and add the following egress rule:
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_vpc_security_group_egress_rule" "allow_outgoing_https" {
+  
+  security_group_id = aws_security_group.unicorn_api_security_group.id
+
+  description = "allow egress via port 443"
+  cidr_ipv4   = "0.0.0.0/0"
+  ip_protocol = "tcp"
+  from_port   = 443
+  to_port     = 443
+}
+```
+
+This [security group egress rule](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_security_group_egress_rule) allows our EC2 instance to make https requests to ECR.
+
+Now that we can connect to ECR, we need to provide access for our EC2 instance to run commands against ECR.
+
+In your `infra` directory create a new file `iam.tf` and add the following:
+
+```terraform
+# /infra/iam.tf
+
+resource "aws_iam_role" "ec2_ecr_readonly_role" {
+  name               = "ec2-ecr-role"
+  assume_role_policy = data.aws_iam_policy_document.ec2_assume_role_policy.json
+
+  inline_policy {
+    name   = "ecr-read-policy"
+    policy = data.aws_iam_policy_document.ec2_container_registry_read_only.json
+  }
+}
+
+data "aws_iam_policy_document" "ec2_assume_role_policy" {
+  statement {
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.amazonaws.com"]
+    }
+  }
+}
+
+data "aws_iam_policy_document" "ec2_container_registry_read_only" {
+  statement {
+    actions = [
+      "ecr:GetAuthorizationToken",
+      "ecr:BatchGetImage",
+      "ecr:GetDownloadUrlForLayer",
+    ]
+    resources = ["*"]
+  }
+}
+```
+
+What we have done is defined an [aws_iam_role](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role) and applied a couple of [aws_iam_policy_documents](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document) which:
+
+1. Define which services can assume this role (`ec2_assume_role_policy`). In this case only EC2 services can assume the role
```

Comment:
**suggestion**: it would be more correct to say that only "the" EC2 service can assume this role.